### PR TITLE
lightning: refactor to reuse in load data, part 3

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #40499

Problem Summary:

### What is changed and how it works?
- move part of code which can be reused into base-encoder（will be used in load data)

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by D3Hunter** [comment]
ci pass https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_integration_br_test/detail/tidb_ghpr_integration_br_test/12491/pipeline

**Comment by D3Hunter** [comment]
@lance6716 @GMHDBJD

File: br/pkg/lightning/backend/kv/base.go [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+	for i := 0; i < len(kvPairs.Pairs); i++ {
+		var encoded [9]byte // The max length of encoded int64 is 9.
+		kvPairs.Pairs[i].RowID = common.EncodeIntRowIDToBuf(encoded[:0], rowID)
+	}
+	e.recordCache = record[:0]
+	return kvPairs, nil
+}
+
+// ProcessColDatum processes the datum of a column.
+func (e *BaseKVEncoder) ProcessColDatum(col *table.Column, rowID int64, inputDatum *types.Datum) (types.Datum, error) {
```
Comment (lance6716) [inline br/pkg/lightning/backend/kv/base.go:197]: Does this need to be public? Seems it only need to be called during `Encode`

**Review by lance6716 — COMMENTED** [review]
9 / 10 files viewed

File: br/pkg/lightning/backend/kv/base.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+		logger:          config.Logger.Logger,
+	}, nil
+}
+
+// GetOrCreateRecord returns a record slice from the cache if possible, otherwise creates a new one.
+func (e *BaseKVEncoder) GetOrCreateRecord() []types.Datum {
+	if e.recordCache != nil {
+		return e.recordCache
+	}
+	return make([]types.Datum, 0, len(e.Columns)+1)
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply br/pkg/lightning/backend/kv/base.go:173]: it's saved after `recode as KV`, didn't change prev logic

File: br/pkg/lightning/backend/kv/base.go (hunk 1 above)
Comment (D3Hunter) [reply br/pkg/lightning/backend/kv/base.go:197]: used here https://github.com/pingcap/tidb/blob/physical-mode/executor/importer/kv_encode.go#L168

**Review by lance6716 — APPROVED** [review]

**Review by gozssky — APPROVED** [review]

File: store/driver/tikv_driver.go [hunk 3]
```diff
@@ -59,6 +60,8 @@ func init() {
 	// Setup the Hooks to dynamic control global resource controller.
 	variable.EnableGlobalResourceControlFunc = tikv.EnableResourceControl
 	variable.DisableGlobalResourceControlFunc = tikv.DisableResourceControl
+	// cannot use this package directly, it causes import cycle
+	importer.GetKVStore = getKVStore
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply store/driver/tikv_driver.go:64]: for load data physical import, importer package will use it to create a kvStore, it's not used in this pr.

[the remark this answers is not in the document]
Comment (gozssky) [reply store/driver/tikv_driver.go:64]: What I mean is how the import cycle is encountered. In other words, is there such a dependency chain: `store/driver` -> A -> B -> .... -> importer?

**Pull request merged**